# Bactiseq: Citations

## [nf-core](https://pubmed.ncbi.nlm.nih.gov/32055031/)

> Ewels PA, Peltzer A, Fillinger S, Patel H, Alneberg J, Wilm A, Garcia MU, Di Tommaso P, Nahnsen S. The nf-core framework for community-curated bioinformatics pipelines. Nat Biotechnol. 2020 Mar;38(3):276-278. doi: 10.1038/s41587-020-0439-x. PubMed PMID: 32055031.

## [Nextflow](https://pubmed.ncbi.nlm.nih.gov/28398311/)

> Di Tommaso P, Chatzou M, Floden EW, Barja PP, Palumbo E, Notredame C. Nextflow enables reproducible computational workflows. Nat Biotechnol. 2017 Apr 11;35(4):316-319. doi: 10.1038/nbt.3820. PubMed PMID: 28398311.

## Pipeline tools

- [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)

> Andrews, S. (2010). FastQC: A Quality Control Tool for High Throughput Sequence Data [Online].

- [Nanoplot](https://academic.oup.com/bioinformatics/article/39/5/btad311/7160911?login=true)
  
> De Coster, W., D’Hert, S., Schultz, D. T., Cruts, M., & Van Broeckhoven, C. (2018). NanoPack: Visualizing and processing long-read sequencing data. Bioinformatics, 34(15), 2666–2669. https://doi.org/10.1093/bioinformatics/bty149

- [Seqkit stats](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0163962)

> Shen, W., Le, S., Li, Y., & Hu, F. (2016). SeqKit: A Cross-Platform and Ultrafast Toolkit for FASTA/Q File Manipulation. PLOS ONE, 11(10), e0163962. https://doi.org/10.1371/journal.pone.0163962

- [GATK4 SamToFastq](https://github.com/broadinstitute/gatk/)

> broadinstitute/gatk: Official code repository for GATK versions 4 and up. (n.d.). Retrieved January 8, 2026, from https://github.com/broadinstitute/gatk/

- [HiFiAdaptFilt](https://pubmed.ncbi.nlm.nih.gov/35193521/)

> Sim, S. B., Corpuz, R. L., Simmonds, T. J., & Geib, S. M. (2022). HiFiAdapterFilt, a memory efficient read processing pipeline, prevents occurrence of adapter sequence in PacBio HiFi reads and their negative impacts on genome assembly. BMC Genomics, 23(1), 157. https://doi.org/10.1186/s12864-022-08375-1

- [Chopper](https://academic.oup.com/bioinformatics/article/34/15/2666/4934939?login=true)

> De Coster, W., D’Hert, S., Schultz, D. T., Cruts, M., & Van Broeckhoven, C. (2018). NanoPack: Visualizing and processing long-read sequencing data. Bioinformatics, 34(15), 2666–2669. https://doi.org/10.1093/bioinformatics/bty149

- [Flye](https://www.nature.com/articles/s41587-019-0072-8)

> Kolmogorov, M., Yuan, J., Lin, Y., & Pevzner, P. A. (2019). Assembly of long, error-prone reads using repeat graphs. Nature Biotechnology, 37(5), 540–546. https://doi.org/10.1038/s41587-019-0072-8

- [Unicycler](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005595)

> Unicycler: Resolving bacterial genome assemblies from short and long sequencing reads | PLOS Computational Biology. (n.d.). Retrieved January 8, 2026, from https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005595

- [SPAdes](https://pmc.ncbi.nlm.nih.gov/articles/PMC3342519/)

> Bankevich, A., Nurk, S., Antipov, D., Gurevich, A. A., Dvorkin, M., Kulikov, A. S., Lesin, V. M., Nikolenko, S. I., Pham, S., Prjibelski, A. D., Pyshkin, A. V., Sirotkin, A. V., Vyahhi, N., Tesler, G., Alekseyev, M. A., & Pevzner, P. A. (2012). SPAdes: A New Genome Assembly Algorithm and Its Applications to Single-Cell Sequencing. Journal of Computational Biology, 19(5), 455–477. https://doi.org/10.1089/cmb.2012.0021

- [Gambit](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0277575)

> GAMBIT (Genomic Approximation Method for Bacterial Identification and Tracking): A methodology to rapidly leverage whole genome sequencing of bacterial isolates for clinical identification | PLOS One. (n.d.). Retrieved January 8, 2026, from https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0277575

- [Kraken2](https://pubmed.ncbi.nlm.nih.gov/31779668/)

> Wood, D. E., Lu, J., & Langmead, B. (2019). Improved metagenomic analysis with Kraken 2. Genome Biology, 20(1), 257. https://doi.org/10.1186/s13059-019-1891-0

- [Pilon](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0112963)

> Walker, B. J., Abeel, T., Shea, T., Priest, M., Abouelliel, A., Sakthikumar, S., Cuomo, C. A., Zeng, Q., Wortman, J., Young, S. K., & Earl, A. M. (2014). Pilon: An Integrated Tool for Comprehensive Microbial Variant Detection and Genome Assembly Improvement. PLOS ONE, 9(11), e112963. https://doi.org/10.1371/journal.pone.0112963

- [Racon](https://pmc.ncbi.nlm.nih.gov/articles/PMC5411768/)

> Vaser, R., Sović, I., Nagarajan, N., & Šikić, M. (2017). Fast and accurate de novo genome assembly from long uncorrected reads. Genome Research, 27(5), 737–746. https://doi.org/10.1101/gr.214270.116

- [Nextpolish](https://academic.oup.com/bioinformatics/article/36/7/2253/5645175)

> Hu, J., Fan, J., Sun, Z., & Liu, S. (2020). NextPolish: A fast and efficient genome polishing tool for long-read assembly. Bioinformatics, 36(7), 2253–2255. https://doi.org/10.1093/bioinformatics/btz891

- [Medaka](https://github.com/nanoporetech/medaka)

> Nanoporetech/medaka. (2026). [Python]. Oxford Nanopore Technologies. https://github.com/nanoporetech/medaka (Original work published 2017)

- [Any2Fasta](https://github.com/tseemann/any2fasta)

> Seemann, T. (2026). Tseemann/any2fasta [Perl]. https://github.com/tseemann/any2fasta (Original work published 2018)

- [CheckM2](https://pubmed.ncbi.nlm.nih.gov/37500759/)

> Chklovski, A., Parks, D. H., Woodcroft, B. J., & Tyson, G. W. (2023). CheckM2: A rapid, scalable and accurate tool for assessing microbial genome quality using machine learning. Nature Methods, 20(8), 1203–1212. https://doi.org/10.1038/s41592-023-01940-w

- [BUSCO](https://academic.oup.com/bioinformatics/article/31/19/3210/211866)

> Simão, F. A., Waterhouse, R. M., Ioannidis, P., Kriventseva, E. V., & Zdobnov, E. M. (2015). BUSCO: Assessing genome assembly and annotation completeness with single-copy orthologs. Bioinformatics, 31(19), 3210–3212. https://doi.org/10.1093/bioinformatics/btv351

- [QUAST](https://pubmed.ncbi.nlm.nih.gov/23422339/)

> Gurevich, A., Saveliev, V., Vyahhi, N., & Tesler, G. (2013). QUAST: Quality assessment tool for genome assemblies. Bioinformatics (Oxford, England), 29(8), 1072–1075. https://doi.org/10.1093/bioinformatics/btt086

- [Bakta](https://pmc.ncbi.nlm.nih.gov/articles/PMC8743544/)

> Schwengers, O., Jelonek, L., Dieckmann, M. A., Beyvers, S., Blom, J., & Goesmann, A. (2021). Bakta: Rapid and standardized annotation of bacterial genomes via alignment-free sequence identification. Microbial Genomics, 7(11), 000685. https://doi.org/10.1099/mgen.0.000685

- [Prokka](https://pmc.ncbi.nlm.nih.gov/articles/PMC8743544/)

> Prokka: Rapid prokaryotic genome annotation | Bioinformatics | Oxford Academic. (n.d.). Retrieved January 8, 2026, from https://academic.oup.com/bioinformatics/article/30/14/2068/2390517

- [Abricate](https://github.com/tseemann/abricate)

> Seemann, T. (2025). Tseemann/abricate [Perl]. https://github.com/tseemann/abricate (Original work published 2014)

- [Mobsuite](https://pmc.ncbi.nlm.nih.gov/articles/PMC6159552/)

> Robertson, J., & Nash, J. H. E. (2018). MOB-suite: Software tools for clustering, reconstruction and typing of plasmids from draft assemblies. Microbial Genomics, 4(8), e000206. https://doi.org/10.1099/mgen.0.000206

- [MLST](https://github.com/tseemann/mlst)

> Seemann, T. (2026). Tseemann/mlst [Perl]. https://github.com/tseemann/mlst (Original work published 2014)

- [CGview](https://academic.oup.com/bioinformatics/article/21/4/537/203470)

> Stothard, P., & Wishart, D. S. (2005). Circular genome visualization and exploration using CGView. Bioinformatics, 21(4), 537–539. https://doi.org/10.1093/bioinformatics/bti054

- [Bandage](https://github.com/rrwick/Bandage)

> Wick, R. (2025). Rrwick/Bandage [C++]. https://github.com/rrwick/Bandage (Original work published 2015)

- [Samtools](https://pmc.ncbi.nlm.nih.gov/articles/PMC2723002/)

> Li, H., Handsaker, B., Wysoker, A., Fennell, T., Ruan, J., Homer, N., Marth, G., Abecasis, G., & Durbin, R. (2009). The Sequence Alignment/Map format and SAMtools. Bioinformatics, 25(16), 2078–2079. https://doi.org/10.1093/bioinformatics/btp352

- [Tinycov](https://github.com/cmdoret/tinycov)

> Matthey-Doret, C. (2025). Cmdoret/tinycov [Python]. https://github.com/cmdoret/tinycov (Original work published 2019)

- [Wget](https://www.gnu.org/software/wget/)

> Wget—GNU Project—Free Software Foundation. (n.d.). Retrieved January 8, 2026, from https://www.gnu.org/software/wget/


## Software packaging/containerisation tools

- [Anaconda](https://anaconda.com)

  > Anaconda Software Distribution. Computer software. Vers. 2-2.4.0. Anaconda, Nov. 2016. Web.

- [Bioconda](https://pubmed.ncbi.nlm.nih.gov/29967506/)

  > Grüning B, Dale R, Sjödin A, Chapman BA, Rowe J, Tomkins-Tinch CH, Valieris R, Köster J; Bioconda Team. Bioconda: sustainable and comprehensive software distribution for the life sciences. Nat Methods. 2018 Jul;15(7):475-476. doi: 10.1038/s41592-018-0046-7. PubMed PMID: 29967506.

- [BioContainers](https://pubmed.ncbi.nlm.nih.gov/28379341/)

  > da Veiga Leprevost F, Grüning B, Aflitos SA, Röst HL, Uszkoreit J, Barsnes H, Vaudel M, Moreno P, Gatto L, Weber J, Bai M, Jimenez RC, Sachsenberg T, Pfeuffer J, Alvarez RV, Griss J, Nesvizhskii AI, Perez-Riverol Y. BioContainers: an open-source and community-driven framework for software standardization. Bioinformatics. 2017 Aug 15;33(16):2580-2582. doi: 10.1093/bioinformatics/btx192. PubMed PMID: 28379341; PubMed Central PMCID: PMC5870671.

- [Docker](https://dl.acm.org/doi/10.5555/2600239.2600241)

  > Merkel, D. (2014). Docker: lightweight linux containers for consistent development and deployment. Linux Journal, 2014(239), 2. doi: 10.5555/2600239.2600241.

- [Singularity](https://pubmed.ncbi.nlm.nih.gov/28494014/)

  > Kurtzer GM, Sochat V, Bauer MW. Singularity: Scientific containers for mobility of compute. PLoS One. 2017 May 11;12(5):e0177459. doi: 10.1371/journal.pone.0177459. eCollection 2017. PubMed PMID: 28494014; PubMed Central PMCID: PMC5426675.
